`pkg/inmemorychannel: TestDispatcher_dispatch` data race

**Describe the bug**

Test `pkg/inmemorychannel: TestDispatcher_dispatch` failed with data race: https://prow.knative.dev/view/gs/knative-prow/pr-logs/pull/knative_eventing/7172/unit-tests_eventing_main/1693891493527293952

```
{Failed  === RUN   TestDispatcher_dispatch
2023-08-22T07:47:18.245Z	DEBUG	inmemorychannel/event_dispatcher_test.go:191	Test servers	{"transformations server": "http://127.0.0.1:36767", "transformations failure server": "http://127.0.0.1:44363", "receiver server": "http://127.0.0.1:39833", "dead letter server": "http://127.0.0.1:44975"}
2023-08-22T07:47:18.245Z	INFO	inmemorychannel/event_dispatcher_test.go:238	Starting dispatcher	{"port": 34963}
2023-08-22T07:47:18.314Z	DEBUG	channel/event_receiver.go:202	Request mapped to channel	{"channel": "svc/channela"}
==================
WARNING: DATA RACE
Read at 0x00c000223ce0 by goroutine 62:
  github.com/cloudevents/sdk-go/v2/event.EventContextV1.GetExtensions()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/event/eventcontext_v1_reader.go:86 +0xae
  github.com/cloudevents/sdk-go/v2/event.(*EventContextV1).GetExtensions()
      <autogenerated>:1 +0x36
  github.com/cloudevents/sdk-go/v2/binding.eventContextToBinaryWriter()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/event_message.go:91 +0x222
  github.com/cloudevents/sdk-go/v2/binding.(*EventMessage).ReadBinary()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/event_message.go:49 +0x7d
  github.com/cloudevents/sdk-go/v2/binding.writeBinary()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/write.go:174 +0x96
  github.com/cloudevents/sdk-go/v2/binding.Write()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/write.go:100 +0x3ba
  github.com/cloudevents/sdk-go/v2/protocol/http.WriteRequest()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/protocol/http/write_request.go:28 +0xd3
  knative.dev/eventing/pkg/kncloudevents.createRequest()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:310 +0x205
  knative.dev/eventing/pkg/kncloudevents.executeRequest()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:246 +0x4ac
  knative.dev/eventing/pkg/kncloudevents.send()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:166 +0x129a
  knative.dev/eventing/pkg/kncloudevents.SendMessage()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:133 +0x26c
  knative.dev/eventing/pkg/kncloudevents.SendEvent()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:116 +0x13b
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).makeFanoutRequest()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:316 +0x213
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).dispatch.func1()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:272 +0x190
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).dispatch.func2()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:274 +0x99

Previous write at 0x00c000223ce0 by goroutine 63:
  runtime.mapassign_faststr()
      /root/.gvm/gos/go1.20.6/src/runtime/map_faststr.go:203 +0x0
  github.com/cloudevents/sdk-go/v2/event.(*EventContextV1).SetExtension()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/event/eventcontext_v1.go:112 +0x1e4
  github.com/cloudevents/sdk-go/v2/binding.(*messageToEventBuilder).SetExtension()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/to_event.go:133 +0xbc
  github.com/cloudevents/sdk-go/v2/binding/transformer.SetExtension.func1()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/transformer/set_metadata.go:41 +0xcd
  github.com/cloudevents/sdk-go/v2/binding.TransformerFunc.Transform()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/transformer.go:24 +0x57
  github.com/cloudevents/sdk-go/v2/binding.Transformers.Transform()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/transformer.go:34 +0xe9
  github.com/cloudevents/sdk-go/v2/binding.(*Transformers).Transform()
      <autogenerated>:1 +0x29
  github.com/cloudevents/sdk-go/v2/binding.Transformers.Transform()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/transformer.go:34 +0xe9
  github.com/cloudevents/sdk-go/v2/binding.(*Transformers).Transform()
      <autogenerated>:1 +0x29
  github.com/cloudevents/sdk-go/v2/binding.Transformers.Transform()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/transformer.go:34 +0x4aa
  github.com/cloudevents/sdk-go/v2/binding.ToEvent()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/to_event.go:40 +0x4ba
  github.com/cloudevents/sdk-go/v2/binding.Write()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/write.go:84 +0x14b
  github.com/cloudevents/sdk-go/v2/protocol/http.WriteRequest()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/protocol/http/write_request.go:28 +0xd3
  knative.dev/eventing/pkg/kncloudevents.createRequest()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:310 +0x205
  knative.dev/eventing/pkg/kncloudevents.executeRequest()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:246 +0x4ac
  knative.dev/eventing/pkg/kncloudevents.send()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:171 +0x161c
  knative.dev/eventing/pkg/kncloudevents.SendMessage()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:133 +0x26c
  knative.dev/eventing/pkg/kncloudevents.SendEvent()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:116 +0x13b
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).makeFanoutRequest()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:316 +0x213
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).dispatch.func1()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:272 +0x190
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).dispatch.func2()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:274 +0x99

Goroutine 62 (running) created at:
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).dispatch()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:271 +0x9a
  knative.dev/eventing/pkg/channel/fanout.createEventReceiverFunction.func2()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:239 +0x217
  knative.dev/eventing/pkg/channel.(*EventReceiver).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/event_receiver.go:221 +0x1191
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:245 +0x64
  knative.dev/eventing/pkg/channel/multichannelfanout.(*EventHandler).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/multichannelfanout/multi_channel_fanout_event_handler.go:139 +0x3ad
  go.opencensus.io/plugin/ochttp.(*Handler).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/vendor/go.opencensus.io/plugin/ochttp/server.go:92 +0x39b
  knative.dev/pkg/network/handlers.(*Drainer).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/vendor/knative.dev/pkg/network/handlers/drain.go:113 +0x1d0
  net/http.serverHandler.ServeHTTP()
      /root/.gvm/gos/go1.20.6/src/net/http/server.go:2936 +0x682
  net/http.(*conn).serve()
      /root/.gvm/gos/go1.20.6/src/net/http/server.go:1995 +0xbd4
  net/http.(*Server).Serve.func3()
      /root/.gvm/gos/go1.20.6/src/net/http/server.go:3089 +0x58

Goroutine 63 (running) created at:
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).dispatch()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:271 +0x9a
  knative.dev/eventing/pkg/channel/fanout.createEventReceiverFunction.func2()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:239 +0x217
  knative.dev/eventing/pkg/channel.(*EventReceiver).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/event_receiver.go:221 +0x1191
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:245 +0x64
  knative.dev/eventing/pkg/channel/multichannelfanout.(*EventHandler).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/multichannelfanout/multi_channel_fanout_event_handler.go:139 +0x3ad
  go.opencensus.io/plugin/ochttp.(*Handler).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/vendor/go.opencensus.io/plugin/ochttp/server.go:92 +0x39b
  knative.dev/pkg/network/handlers.(*Drainer).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/vendor/knative.dev/pkg/network/handlers/drain.go:113 +0x1d0
  net/http.serverHandler.ServeHTTP()
      /root/.gvm/gos/go1.20.6/src/net/http/server.go:2936 +0x682
  net/http.(*conn).serve()
      /root/.gvm/gos/go1.20.6/src/net/http/server.go:1995 +0xbd4
  net/http.(*Server).Serve.func3()
      /root/.gvm/gos/go1.20.6/src/net/http/server.go:3089 +0x58
==================
==================
WARNING: DATA RACE
Read at 0x00c0003f08a8 by goroutine 62:
  github.com/cloudevents/sdk-go/v2/event.EventContextV1.GetExtensions()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/event/eventcontext_v1_reader.go:91 +0x1b9
  github.com/cloudevents/sdk-go/v2/event.(*EventContextV1).GetExtensions()
      <autogenerated>:1 +0x36
  github.com/cloudevents/sdk-go/v2/binding.eventContextToBinaryWriter()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/event_message.go:91 +0x222
  github.com/cloudevents/sdk-go/v2/binding.(*EventMessage).ReadBinary()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/event_message.go:49 +0x7d
  github.com/cloudevents/sdk-go/v2/binding.writeBinary()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/write.go:174 +0x96
  github.com/cloudevents/sdk-go/v2/binding.Write()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/write.go:100 +0x3ba
  github.com/cloudevents/sdk-go/v2/protocol/http.WriteRequest()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/protocol/http/write_request.go:28 +0xd3
  knative.dev/eventing/pkg/kncloudevents.createRequest()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:310 +0x205
  knative.dev/eventing/pkg/kncloudevents.executeRequest()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:246 +0x4ac
  knative.dev/eventing/pkg/kncloudevents.send()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:166 +0x129a
  knative.dev/eventing/pkg/kncloudevents.SendMessage()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:133 +0x26c
  knative.dev/eventing/pkg/kncloudevents.SendEvent()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:116 +0x13b
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).makeFanoutRequest()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:316 +0x213
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).dispatch.func1()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:272 +0x190
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).dispatch.func2()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:274 +0x99

Previous write at 0x00c0003f08a8 by goroutine 63:
  github.com/cloudevents/sdk-go/v2/event.(*EventContextV1).SetExtension()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/event/eventcontext_v1.go:112 +0x1f0
  github.com/cloudevents/sdk-go/v2/binding.(*messageToEventBuilder).SetExtension()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/to_event.go:133 +0xbc
  github.com/cloudevents/sdk-go/v2/binding/transformer.SetExtension.func1()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/transformer/set_metadata.go:41 +0xcd
  github.com/cloudevents/sdk-go/v2/binding.TransformerFunc.Transform()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/transformer.go:24 +0x57
  github.com/cloudevents/sdk-go/v2/binding.Transformers.Transform()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/transformer.go:34 +0xe9
  github.com/cloudevents/sdk-go/v2/binding.(*Transformers).Transform()
      <autogenerated>:1 +0x29
  github.com/cloudevents/sdk-go/v2/binding.Transformers.Transform()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/transformer.go:34 +0xe9
  github.com/cloudevents/sdk-go/v2/binding.(*Transformers).Transform()
      <autogenerated>:1 +0x29
  github.com/cloudevents/sdk-go/v2/binding.Transformers.Transform()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/transformer.go:34 +0x4aa
  github.com/cloudevents/sdk-go/v2/binding.ToEvent()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/to_event.go:40 +0x4ba
  github.com/cloudevents/sdk-go/v2/binding.Write()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/binding/write.go:84 +0x14b
  github.com/cloudevents/sdk-go/v2/protocol/http.WriteRequest()
      /home/prow/go/src/knative.dev/eventing/vendor/github.com/cloudevents/sdk-go/v2/protocol/http/write_request.go:28 +0xd3
  knative.dev/eventing/pkg/kncloudevents.createRequest()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:310 +0x205
  knative.dev/eventing/pkg/kncloudevents.executeRequest()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:246 +0x4ac
  knative.dev/eventing/pkg/kncloudevents.send()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:171 +0x161c
  knative.dev/eventing/pkg/kncloudevents.SendMessage()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:133 +0x26c
  knative.dev/eventing/pkg/kncloudevents.SendEvent()
      /home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher.go:116 +0x13b
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).makeFanoutRequest()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:316 +0x213
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).dispatch.func1()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:272 +0x190
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).dispatch.func2()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:274 +0x99

Goroutine 62 (running) created at:
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).dispatch()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:271 +0x9a
  knative.dev/eventing/pkg/channel/fanout.createEventReceiverFunction.func2()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:239 +0x217
  knative.dev/eventing/pkg/channel.(*EventReceiver).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/event_receiver.go:221 +0x1191
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:245 +0x64
  knative.dev/eventing/pkg/channel/multichannelfanout.(*EventHandler).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/multichannelfanout/multi_channel_fanout_event_handler.go:139 +0x3ad
  go.opencensus.io/plugin/ochttp.(*Handler).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/vendor/go.opencensus.io/plugin/ochttp/server.go:92 +0x39b
  knative.dev/pkg/network/handlers.(*Drainer).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/vendor/knative.dev/pkg/network/handlers/drain.go:113 +0x1d0
  net/http.serverHandler.ServeHTTP()
      /root/.gvm/gos/go1.20.6/src/net/http/server.go:2936 +0x682
  net/http.(*conn).serve()
      /root/.gvm/gos/go1.20.6/src/net/http/server.go:1995 +0xbd4
  net/http.(*Server).Serve.func3()
      /root/.gvm/gos/go1.20.6/src/net/http/server.go:3089 +0x58

Goroutine 63 (running) created at:
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).dispatch()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:271 +0x9a
  knative.dev/eventing/pkg/channel/fanout.createEventReceiverFunction.func2()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:239 +0x217
  knative.dev/eventing/pkg/channel.(*EventReceiver).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/event_receiver.go:221 +0x1191
  knative.dev/eventing/pkg/channel/fanout.(*FanoutEventHandler).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/fanout/fanout_event_handler.go:245 +0x64
  knative.dev/eventing/pkg/channel/multichannelfanout.(*EventHandler).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/pkg/channel/multichannelfanout/multi_channel_fanout_event_handler.go:139 +0x3ad
  go.opencensus.io/plugin/ochttp.(*Handler).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/vendor/go.opencensus.io/plugin/ochttp/server.go:92 +0x39b
  knative.dev/pkg/network/handlers.(*Drainer).ServeHTTP()
      /home/prow/go/src/knative.dev/eventing/vendor/knative.dev/pkg/network/handlers/drain.go:113 +0x1d0
  net/http.serverHandler.ServeHTTP()
      /root/.gvm/gos/go1.20.6/src/net/http/server.go:2936 +0x682
  net/http.(*conn).serve()
      /root/.gvm/gos/go1.20.6/src/net/http/server.go:1995 +0xbd4
  net/http.(*Server).Serve.func3()
      /root/.gvm/gos/go1.20.6/src/net/http/server.go:3089 +0x58
==================
2023-08-22T07:47:18.323Z	DEBUG	channel/event_receiver.go:202	Request mapped to channel	{"channel": "svc/channelb"}
2023/08/22 07:47:18 failed to send the request: Post "http://zipkin.zipkin.svc.cluster.local:9411/api/v2/spans": dial tcp: lookup zipkin.zipkin.svc.cluster.local: no such host
    testing.go:1446: race detected during execution of test
--- FAIL: TestDispatcher_dispatch (0.17s)
}
```

 **Expected behavior**

No data race errors

 **To Reproduce**

Run the test `pkg/inmemorychannel: TestDispatcher_dispatch`

 **Knative release version**

`devel`

 **Additional context**

/priority important-soon

--- a/pkg/kncloudevents/event_dispatcher.go
+++ b/pkg/kncloudevents/event_dispatcher.go
@@ -111,7 +111,11 @@ type senderConfig struct {
 
 // SendEvent sends the given event to the given destination.
 func SendEvent(ctx context.Context, event event.Event, destination duckv1.Addressable, options ...SendOption) (*DispatchInfo, error) {
-	message := binding.ToMessage(&event)
+	// clone the event since:
+	// - we mutate the event and the callers might not expect this
+	// - it might produce data races if the caller is trying to read the event in different go routines
+	c := event.Clone()
+	message := binding.ToMessage(&c)
 
 	return SendMessage(ctx, message, destination, options...)
 }

